securebanking-openbanking-uk-gateway-conversion-filter/src/main/java/com/forgerock/securebanking/uk/gateway/conversion/filter/IntentConverterFilter.java

```diff
@@ -0,0 +1,261 @@
+
+import static org.forgerock.http.protocol.Entity.APPLICATION_JSON_CHARSET_UTF_8;
+import static org.forgerock.json.JsonValueFunctions.enumConstant;
+import static org.forgerock.openig.el.Bindings.bindings;
+import static org.forgerock.openig.util.AsyncFunctions.asyncResultHandler;
+import static org.forgerock.util.promise.NeverThrowsException.neverThrown;
+import static org.forgerock.util.promise.Promises.newResultPromise;
+
+/**
+ * Filter to convert IDM json intent objects to OB data model objects.
+ *
+ * This filter must have received {@code intentType} as required to identify the intent type {@link IntentType#toString()} to instance the converter<br/>
+ * Failures from the `Converter instance` is either a {@link RuntimeException} that will catch to build a {@link ResponseException}.
+ *
+ * Configuration options:
+ *
+ * <pre>
+ * {@code {
+ *      "name": "IntentConverterFilter-name"
+ *      "type": "IntentConverterFilter",
+ *      "config": {
+ *         "messageType"  MessageType              [REQUIRED - The type of message for which to convert the entity, Must be either "REQUEST" or "RESPONSE"]
+ *         "entity"       Expression<String>       [OPTIONAL - A jsonPayload content to be converted in string format. Default : not replaced.]
+ *         "resultTo"     List<MessageType>        [OPTIONAL - Indicates where will set the conversion result. Must be either "REQUEST" or "RESPONSE". Default REQUEST.]
+ *      }
+ *  }
+ *  }
+ * </pre>
+ * <p>Example</p>
+ * <pre>
+ * {@code {
+ *      "name": "IntentConverterFilter-name"
+ *      "type": "IntentConverterFilter",
+ *      "config": {
+ *         "messageType: "request",
+ *         "entity": "#{request.entity.string}",
+ *         "resultTo": ["REQUEST", "RESPONSE],
+ *      }
+ *  }
+ *  }
+ * </pre>
+ */
+public class IntentConverterFilter implements Filter {
+
+    private static final Logger logger = LoggerFactory.getLogger(IntentConverterFilter.class);
+
+    private final MessageType messageType;
+    private final List<MessageType> resultTo;
+    private final Expression<String> entity;
+    private static final ObjectMapper MAPPER = GenericConverterMapper.getMapper();
+
+    /**
+     * Constructor
+     * @param messageType
+     * @param entity
+     * @param resultTo
+     */
+    public IntentConverterFilter(final MessageType messageType, final Expression<String> entity, final List<MessageType> resultTo) {
+        this.messageType = messageType;
+        this.entity = entity;
+        this.resultTo = resultTo;
+    }
+
+    @Override
+    public Promise<Response, NeverThrowsException> filter(Context context, Request request, Handler next) {
+        if (logger.isInfoEnabled()) {
+            printFilterInfo();
+        }
+
+        if (messageType.equals(MessageType.REQUEST)) {
+            return next.handle(context, request).thenAsync(response -> processRequest(request, response, bindings(context, request)));
+        }
```

Comment:
I think the request needs converting before it is handed on to the rest of the chain, doing work in the thenAsync is on the response chain which means the request will have been passed on to the other filters (and to the RS) unconverted.

Something along the lines of:
```
if (messageType.equals(MessageType.REQUEST)) {
            processRequest(request, null, bindings(context, request));
            return next.handle(context, request);
        }
```

I think the processRequest method sig needs reviewing, the request should not be concerned with the response, the filter should have the following 3 modes:

-  Convert requests only, do the conversion on the request entity and hand the updated entity on to the next handler in the chain
- Converts response only, do the conversion on the response entity and hand the updated entity on to the next handler
- Convert Request entity and Response entity (use both pieces of logic from above)